core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSourceCDCSupport.scala

```diff
@@ -148,13 +148,31 @@ trait DeltaSourceCDCSupport { self: DeltaSource =>
           // We allow entries with no file actions and index as [[DeltaSourceOffset.BASE_INDEX]]
           // that are used primarily to update latest offset when no other
           // file action based entries are present.
-          fileActions.filter(indexedFile => hasAddsOrRemoves(indexedFile) ||
-            hasNoFileActionAndStartIndex(indexedFile))
-            .filter(
-              isValidIndexedFile(_, fromVersion, fromIndex, endOffset)
-            ).takeWhile { indexedFile =>
-            admissionControl.admit(Option(indexedFile.getFileAction))
-          }.toIterator
+          val filteredFiles = fileActions
+            .filter { indexedFile =>
+              hasAddsOrRemoves(indexedFile) || hasNoFileActionAndStartIndex(indexedFile)
+            }
+            .filter(isValidIndexedFile(_, fromVersion, fromIndex, endOffset))
+          val filteredFileActions = filteredFiles.flatMap(f => Option(f.getFileAction))
```

Comment:
This `flatMap(... => Option(...))` construct makes it hard(er) to get the gist of the change. Why not simply `filter(_.getFileAction != null)`?

Looks clear(er) and less objects put on the JVM stack.